presto-main/src/main/java/com/facebook/presto/cost/CachingStatsProvider.java

```diff
@@ -42,6 +44,7 @@
     private final Session session;
     private final TypeProvider types;
 
+    private final PlanHasher planHasher = new CachingPlanHasher();
```

Comment:
Meta question: do we really need to tunnel this everywhere? Or we can inject a singleton `CachingPlanHasher` and pass it through `HistoryBasedPlanStatisticsCalculator`'s constructor?

Reply:
If we make it a singleton, we will have to take care of when to drop items from cache. Right now, this object is cleared after planning is done. 

For singleton, we will either have to manually remove them(which is ugly), or have a cache which drops plan nodes after 5 mins.

Reply:
Either you could have a 5-min TTL with a max size. Or if you wanna do it more elegantly, you can create your own cache class with an loading cache implementation (which is strongly recommended than maintaining your own hashmap) and put `QueryManager` there. Check how `QueryManager` is used in many other places. Then you can have a scheduled executor to purged dead queries.

Reply:
I'll go with 5 min TTL for now and put a todo, mostly because sapphire doesn't have a query manager implementation.